include/bout/indexoffset.hxx

```diff
@@ -0,0 +1,99 @@
+#ifndef __INDEXOFFSET_H__
+#define __INDEXOFFSET_H__
+
+#include "bout/mesh.hxx"
+#include "bout/region.hxx"
+
+/// Helper class for offsetting Ind2D/Ind3D
+///
+/// Provides methods for offsetting by fixed amounts in x, y, z, as
+/// well as a generic method for offsetting by any amount in multiple
+/// directions
+///
+/// Also provides helper methods for converting Ind2D/Ind3D to x, y, z
+/// indices
+///
+/// Examples
+/// --------
+///
+///     Field3D field, result;
+///     auto index = std::begin(region);
+///     IndexOffset<Ind3D> offset(*mesh);
+///
+///     result = field[offset.yp(*index)] - field[offset.ym(*index)];
+///
+template <typename T = Ind3D>
+struct IndexOffset {
+
+  static_assert(std::is_base_of<Ind2D, T>::value || std::is_base_of<Ind3D, T>::value,
+                "IndexOffset must be templated with either Ind2D or Ind3D");
+
+  Mesh &mesh;
+  const int nx, ny, nz;
+
+  IndexOffset(Mesh &mesh)
+      : mesh(mesh), nx(mesh.LocalNx), ny(mesh.LocalNy),
+        nz(std::is_base_of<Ind3D, T>::value ? mesh.LocalNz : 1) {}
+
+  /// Convenience functions for converting \p index to (x, y, z)
+  int x(T index) const { return (index.ind / nz) / ny; }
+  int y(T index) const { return (index.ind / nz) % ny; }
+  int z(T index) const { return (index.ind % nz); }
+
+  /// Positive offset \p index by \p dx in x, default to offset by one
+  const inline T xp(T index, int dx = 1) const {
+#if CHECK > 3
+    if (x(index) + dx < 0 or x(index) + dx >= nx) {
+      throw BoutException("Offset in x (%d) would go out of bounds at %d", dx, index.ind);
+    }
+#endif
+    return index + (dx * ny * nz);
+  }
+  /// Negative offset \p index by \p dx in x, default to offset by one
+  const inline T xm(T index, int dx = 1) const { return xp(index, -dx); }
+  /// Positive offset \p index by \p dy in y, default to offset by one
+  const inline T yp(T index, int dy = 1) const {
+#if CHECK > 3
+    if (y(index) + dy < 0 or y(index) + dy >= ny) {
+      throw BoutException("Offset in y (%d) would go out of bounds at %d", dy, index.ind);
+    }
+#endif
+    return index + (dy * nz);
+  }
+  /// Negative offset \p index by \p dy in y, default to offset by one
+  const inline T ym(T index, int dy = 1) const { return yp(index, -dy); }
+  /// Positive offset \p index by \p dz in z, default to offset by one.
+  /// Periodic in z, cannot handle negative offsets
+  const inline T zp(T index, int dz = 1) const {
+    ASSERT3(dz > 0);
+    if (nz == 1) {
```

Comment:
Does branching here and in `zm` have performance implications?

Reply:
Plausibly, but I suspect the branch predictor will save us. It will destroy any chance of vectorising loops, which might be the bigger deal. I'm not sure what can be done to remove it -- the offsetting can't be done for `Field2D`, and it's not possible to specialise a single member function.

Reply:
What is the expect behaviour when using `Field2D`, ie `nz == 1`?  I'd expect `zp` to be ignored, in which case `zp(index2d, dz)` should return `index2d` (the `nz==1` logic), not `index2d - 1 + dz` (the first branch of the ternary).

Reply:
With the implicit assumption that `dz<=nz` we have `index2D == index2D -1 + dz`